modin/backends/pandas/query_compiler.py

```diff
@@ -2576,34 +2576,71 @@ def groupby_agg(
         if callable(agg_func):
             agg_func = wrap_udf_function(agg_func)
 
-        if is_multi_by:
-            return super().groupby_agg(
-                by=by,
-                is_multi_by=is_multi_by,
-                axis=axis,
-                agg_func=agg_func,
-                agg_args=agg_args,
-                agg_kwargs=agg_kwargs,
-                groupby_kwargs=groupby_kwargs,
-                drop=drop,
-            )
-
-        by = by.to_pandas().squeeze() if isinstance(by, type(self)) else by
-
         # since we're going to modify `groupby_kwargs` dict in a `groupby_agg_builder`,
         # we want to copy it to not propagate these changes into source dict, in case
         # of unsuccessful end of function
         groupby_kwargs = groupby_kwargs.copy()
 
         as_index = groupby_kwargs.get("as_index", True)
 
-        def groupby_agg_builder(df):
+        if isinstance(by, type(self)):
+            internal_by = by.columns
+            by = [by]
+        else:
+            if not isinstance(by, list):
+                by = [by]
+            internal_by = [o for o in by if isinstance(o, str) and o in self.columns]
+            internal_qc = (
+                [self.getitem_column_array(internal_by)] if len(internal_by) else []
+            )
+
+            by = internal_qc + by[len(internal_by) :]
+
+        broadcastable_by = [o._modin_frame for o in by if isinstance(o, type(self))]
+        not_broadcastable_by = [o for o in by if not isinstance(o, type(self))]
+
+        def groupby_agg_builder(df, by=None, drop=False):
```

Comment:
Can you provide some overview on `groupby_agg_builder` functionality? What is difference between old and new versions?

Reply:
Here is a reproduction of `grouoby_agg_builder` work:
- `df` contains partition on what we want to apply groupby
- `by` is our `broadcastable_by`
- `internal_by` is a list of 'by's that supposed to be from `df`
```python
 df:                 | by:			| internal_by:
  col0  col1  col2   |   col0 col3 __reduced__  |   ["col0", "col3"]
0    a     1     5   | 0    a  one           x  |
1    b     2     6   | 1    b  one           y  |
2    a     3     7   | 2    a  two           x  |
3    a     4     8   | 3    a  two           y  |
```

First, we need to decide how we will process groupby under `df`, here is the rule that `groupby_agg_builder` follows in case of multi 'by':
We must have all `internal_by` columns inside `df` and then group by these column names.

So firstly we're splitting `by` into:
```python
 internal_by_df:  | external_by_df:
  col0 col3       |   __reduced__
0    a  one       | 0           x
1    b  one       | 1           y
2    a  two       | 2           x
3    a  two       | 3           y
```

Then we're appending 'by' columns that mismatched in `df` from `internal_by_df`
```python
>>> df
  col0  col1  col2 col3
0    a     1     5  one
1    b     2     6  one
2    a     3     7  two
3    a     4     8  two
```

Next step is to provide valid 'by' to groupby on `df`, so:
```python
>>> valid_by = list(internal_by_df.columns) + [split_df_into_a_list_of_series(external_by_df)] + not_broadcastable_by
>>> valid_by
['col0', 'col3', 
0    x
1    y
2    x
3    y
Name: __reduced__, dtype: object]
```

Then we're just applying groupby with `as_index=True` to track index-metadata
```python
>>> grouped_df = df.groupby(by=by, axis=axis, **groupby_kwargs)
>>> result = agg_func(grouped_df, **agg_kwargs)
>>> result
                       col1  col2
col0 col3 __reduced__
a    one  x               1     5
     two  x               3     7
          y               4     8
b    one  y               2     6
```

Then we have two possible ways depends on `as_index` parameter, if it's True, then we're ready to return our result,
else we need to insert back our `internal_by` columns to the result, all we have to do is drop all level from index whose name is `None` or `"__reduced__"` and then just do `reset_index`. However, sometimes we may have a naming conflict between index levels and columns. The thing is that sometimes pandas groupby does not drop 'by' columns from the resulted frame (for example in case of `apply`)
```python
>>> df
  col0  col1  col2 col3
0    a     1     5  one
1    b     2     6  one
2    a     3     7  two
3    a     4     8  two
>>> df.groupby(["col0", "col3"]).apply(sum)
          col0  col1  col2    col3
col0 col3
a    one     a     1     5     one
     two    aa     7    15  twotwo
b    one     b     2     6     one
```
In that case, we must resolve naming conflicts similarly with pandas:
- If 'by' columns contains at least one categorical column, then we're removing duplicated names from results columns [[1]](https://github.com/pandas-dev/pandas/blob/61f67b63ee8c01dad3858c3a01c34839d131d041/pandas/core/groupby/groupby.py#L2797)
- Otherwise, we're deleting duplicated names from index levels [[1]](https://github.com/pandas-dev/pandas/blob/61f67b63ee8c01dad3858c3a01c34839d131d041/pandas/core/groupby/generic.py#L1614-L1625)

And this is exactly what does [this](https://github.com/dchigarev/modin/blob/ed8d79f53cde7bdff495bfc08a96a7ee21ff0098/modin/backends/pandas/query_compiler.py#L2666-L2695) piece of code in `groupby_agg_builder`

Reply:
Previously, `groupby_agg_builder` only has logic about grouping on single-column 'by' and applying aggregation function to a groupby object

Reply:
Thanks for descriptive explanation!